### DISCUSSION TITLE
On plugin maintenance

### DISCUSSION THREAD

**Discussion by introt** [zim-desktop-wiki/zim-desktop-wiki#1936] [Development]
**Content warnings:** Stray thoughts, references to bureaucracy, suggestions better suited to larger projects

Having delved into plugin development, I've started wondering whether it would make sense to split up some of the plugins into their own repositories under the zim-desktop-wiki org and using submodules to pin stable versions to include in releases. This would give plugin authors the necessary authority (pun intended) to solve issues and PRs while keeping the main repo/releases under Jaap's control.

Good examples of plugins to submodule include ones that started off as third-party plugins (eg. the [Command palette](https://github.com/datenwurm/zim-plugin-command-palette)) and are unlikely to require (further) API modifications. Contrast this with eg. the ToC plugin and #1626 (though said API changes could be included in the PR for updating the submodule commit).

Some first-party plugins (like table editor and linkmap) gather a disproportionate amount of tickets, but splitting them likely wouldn't help much, as more decks doesn't equal more hands in total across aforementioned decks. Instead, would listing all those issues at a single place on a project board make them more or less attractive for a prospective dev?

**Comment by introt** [comment]
Could also pave way for a plugin downloader plugin (with git as backend) to make plugin discovery & dev feature testing easier for end users, while keeping things secure by requiring commits signed by Jaap by default etc

**Comment by jaap-karssenberg** [comment]
My idea was to set up a template github repository for plugins where the
layout is such that it can be checkout directly in the zim plugin folder.
This way updating etc. can be done using git commands. Benefit of avoiding
lot of "downloader" code and tie in with an existing development
infrastructure.

Together with a central hosted list of plugin repositories you will have
all you need to show this in the application.

-- Jaap

**Reply by introt** [reply]
I was aiming to build on that idea by introducing using the GitHub org & REST API, but the more I think about it the less it makes sense compared to just having the plugin authors send in PRs to the central repo with commit hashes/checksums to prevent foul play.

**Reply by introt** [reply]
Here's a very minimal template repo: https://github.com/introt/zim-plugin-template

What are the minimum reqs for moving this under the org account? Imo it should contain boilerplate for tests etc.

**Comment by jaap-karssenberg** [comment]
Happy to move it under the org account - just need to check out how to do
that - probably I can clone it as the org ?

**Reply by introt** [reply]
Found "transfer" under the repo settings, so consider it as done :)

How'd you like to handle "flavors" (like `minimal`, `ImageGeneratorPlugin` etc)? I'd have gone for branches in a single repo, but that doesn't seem to be how the Template Repository functionality is build - and might be confusing for git newbies. On the other hand, having a repo per template may make updating them more of a chore.

**Comment by jaap-karssenberg** [comment]
Unless you foresee many flavors I would just put those in a single init
file and use comments to guide people what to delete and what to keep.

Other thought is that we should think of a mechanism to provide user
documentation in such a plugin repo and make it visible in the application.

- Jaap

Op wo 25 mei 2022 14:18 schreef introt ***@***.***>:

**Reply by introt** [reply]
Sounds like a plan - boilerplate, examples etc. can then live in separate modules (files) that can be edited or deleted as required.

I think adding a `plugin_manual` attribute to `PluginClass` to complement the existing `plugin_info` would be the way to go - it'd also allow translations via `_`. As with `README`s and `setup.py`s, the manual page source wouldn't necessarily need to live in the same file nor a single file - one could also use `@property` to do all sorts of dynamic stuff (think refresh on reload). The necessary index hooks would also pave way for other "virtual"/dynamically generated pages, and could be exposed via the plugin API.

**Comment by fmos** [comment]
Was wondering whether you were considering `git submodule` for plugins, found this thread und just used the template for a new plugin: https://github.com/fmos/inlinesuggestions

If and when I consider this plugin as ready for mainstream (has tests etc.), I will file a PR for a submodule under plugins.

**Reply by introt** [reply]
Afaik we haven't made decisions regarding submodules, but you can throw a PR at the wall and see if it sticks; the template is part of the "plugin downloader plugin" - feel free to improve it with some test examples etc if you find the time!
